crates/common/src/config.rs

```diff
@@ -34,73 +67,214 @@ pub fn safe_default<V: Debug + Eq>(opt: Option<V>, default: V) -> anyhow::Result
+///
+/// # Behavior
+///
+/// 1. Computes a `system_config_hash` from the system configuration settings in `rollup_config.genesis`.
+///    If the system configuration is absent, a default zeroed 32-byte array is used.
+/// 2. Serializes various fields of `RollupConfig`, including genesis information, block time settings,
+///    protocol parameters, hardfork timings, and address-specific fields. These fields are concatenated
+///    into a single byte array.
+/// 3. The resulting byte array is hashed using SHA-256 to produce a 32-byte digest.
+/// 4. Returns the computed hash if all operations succeed.
+///
+/// # Notes
+///
+/// * `safe_default` is used extensively to provide fallback values for optional configuration
+///   fields, ensuring robust handling of missing or invalid data.
+/// * All numeric values are serialized in big-endian format for consistency.
 pub fn config_hash(rollup_config: &RollupConfig) -> anyhow::Result<[u8; 32]> {
-    // todo: check whether we need to include this, or if it is loaded from the config address
-    let system_config_hash: [u8; 32] = rollup_config
-        .genesis
-        .system_config
-        .as_ref()
-        .map(|system_config| {
-            let fields = [
-                system_config.batcher_address.0.as_slice(),
-                system_config.overhead.to_be_bytes::<32>().as_slice(),
-                system_config.scalar.to_be_bytes::<32>().as_slice(),
-                system_config.gas_limit.to_be_bytes().as_slice(),
-                safe_default(system_config.base_fee_scalar, u64::MAX)
-                    .context("base_fee_scalar")?
-                    .to_be_bytes()
-                    .as_slice(),
-                safe_default(system_config.blob_base_fee_scalar, u64::MAX)
-                    .context("blob_base_fee_scalar")?
-                    .to_be_bytes()
-                    .as_slice(),
-                safe_default(system_config.eip1559_denominator, u32::MAX)
-                    .context("eip1559_denominator")?
-                    .to_be_bytes()
-                    .as_slice(),
-                safe_default(system_config.eip1559_elasticity, u32::MAX)
-                    .context("eip1559_elasticity")?
-                    .to_be_bytes()
-                    .as_slice(),
-            ]
-            .concat();
-            let digest = SHA2::hash_bytes(fields.as_slice());
-
-            Ok::<[u8; 32], anyhow::Error>(digest.as_bytes().try_into()?)
-        })
-        .unwrap_or(Ok([0u8; 32]))?;
     let rollup_config_bytes = [
+        // genesis
         rollup_config.genesis.l1.hash.0.as_slice(),
         rollup_config.genesis.l1.number.to_be_bytes().as_slice(),
         rollup_config.genesis.l2.hash.0.as_slice(),
         rollup_config.genesis.l2.number.to_be_bytes().as_slice(),
         rollup_config.genesis.l2_time.to_be_bytes().as_slice(),
-        system_config_hash.as_slice(),
+        safe_default(
```

Comment:
Especially for longer structs this manual approach seems rather error-prone, consider using a derived serializer to automate this.